Add secondary subtitle support to html video player

is343 opened the pull request:
<!--
Ensure your title is short, descriptive, and in the imperative mood (Fix X, Change Y, instead of Fixed X, Changed Y).
For a good inspiration of what to write in commit messages and PRs please review https://chris.beams.io/posts/git-commit/ and our https://docs.jellyfin.org/general/contributing/issues.html page.
-->


This PR adds the ability to view multiple subtitle tracks via the html video player as requested in this [feature request](https://features.jellyfin.org/posts/936/multi-language-subtitles).
I have done this by allowing the user to select a secondary track when a primary subtitle track is already playing. Depending on browser support, it will play natively or use the custom subtitle renderer.
The selected secondary track index is saved in the player instance's state, so the selected secondary subtitle will be remembered for the current play session. This is convenient for watching several episodes of a show, for example. The secondary track index is not saved anywhere else, so it will need to be set again when coming back for a new session. Subtitle offsets will also apply to both tracks. I also applied a fix for a firefox bug where the current visible track would not clear when setting an offset.

Since we are only handling the secondary subtitles on the client-side, only `External` text-based subtitles are currently supported. Anything burned in or done on the server-side will require changes there before we can make them work.

I was originally planning on allowing the user to select any number of extra subtitles by handling arrays of indexes, but that seemed like a lot more work to build in such a way that would not break current compatibility with existing methods or plugins. Also, I don't see the number of users wanting more than 2 subtitles to be that many.

**Changes**
<!-- Describe your changes here in 1-5 sentences. -->
- ~~fix issue with opening multiple actionsheets~~ #3900
  - ~~required since I added a new nested actionsheet~~
- add actionsheet for selecting secondary subtitles
  - All of the other actionsheets that are opened within another actionsheet do not have titles
  - I added a title to the secondary subtitle actionsheet because the contents of this actionsheet are likely to be the same as the one we just came from (main subtitle actionsheet). It might be easy for the user to not notice that they are on a new menu. So, I added the title to this one to let the user know what they are looking at.
- added new `playbackmanager` methods for handling secondary subtitles, getters, setters
  - resetting secondary subtitle state in `setSubtitleStreamIndex` when primary subtitle is cleared
- added new methods and updated logic to the main html player plugin
  - added logic to make sure we know which subtitle to manage (create, update, destroy) by adding a new argument for the target track index
  - updated logic for subtitle offsets. There was an issue when passing the offset to the second track. It was not getting the correct relative offset. Also no longer looping through all cues if the calculated offset === 0
  - added a workaround to fixed a bug in firefox where the currently visible text track would not clear when offset was changed
  - add logic to fetch the currently selected secondary track and start playing it
- added a new css class for the custom rendered secondary subtitles.
  - using `!important` because the `min-height` is set to over `4em` directly to the element's styles. This makes the spacing between the subtitles too big. Also for used in `margin-bottom` just to give a little bit of space between the two, while still being able to use `margin: auto`.

**Issues**
<!-- Tag any issues that this PR solves here.
ex. Fixes # -->
[multi-language-subtitles feature request](https://features.jellyfin.org/posts/936/multi-language-subtitles)


Sorry for the bad quality gifs. I had to keep them under 10mb to add them to this PR.

Chrome - native subtitles
![Screen Recording 2022-09-14 at 5 47 24 PM](https://user-images.githubusercontent.com/30599893/190533966-ef69e951-6b54-4504-b1a5-bb0a82782be0.gif)

Chrome - custom renderer - tested by setting `requiresCustomSubtitlesElement` return `true`. (this was before I updated the spacing in the custom renderer css)
![Screen Recording 2022-09-14 at 5 40 28 PM](https://user-images.githubusercontent.com/30599893/190534100-55950ea4-8fe8-4f6f-8fd4-0044aa3713dc.gif)

(with current updated spacing for custom rendered subtitles)
<img width="905" alt="Screen Shot 2022-09-15 at 9 33 23 PM" src="https://user-images.githubusercontent.com/30599893/190537397-0c186b3b-1c92-4b08-871c-53ee8d69078c.png">


Firefox - native
![Screen Recording 2022-09-14 at 6 08 24 PM](https://user-images.githubusercontent.com/30599893/190534251-1fcb3927-4a73-40c5-9133-0739c1b07b2a.gif)


Firefox - (before fix) showing the bug where the current visible track does not clear when changing the offset. You can see the currently visible text (music notes) remains rendered until it reaches the point where it would be rendered again with the new offset timing
![Screen Recording 2022-09-14 at 6 10 24 PM](https://user-images.githubusercontent.com/30599893/190534347-20ef9269-afa3-43c6-ad08-a5b2bfd74b26.gif)

Firefox - (after fix) no longer showing the the bug from above
![Screen Recording 2022-09-14 at 8 01 33 PM](https://user-images.githubusercontent.com/30599893/190534674-3506ce7c-82dd-41da-9e6b-65b37fbfc7dc.gif)

dmitrylyzo reviewed (commented):
I think we should only allow a combination of text-based subtitles.
At the moment, ASS+ASS obviously doesn't work - it requires a second instance of JSO, but makes no sense since they will overlap.
SRT+ASS or ASS+SRT are possible, but sometimes ASS is turned off when changing the subtitles. But again, this will look weird.

is343 commented on src/plugins/htmlVideoPlayer/plugin.js:
```diff
@@ -1510,6 +1685,7 @@ function tryRemoveElement(elem) {
 
         list.push('SetBrightness');
         list.push('SetAspectRatio');
+        list.push('SecondarySubtitles');
```
WebOS should be using the custom css subtitle rendering. There is a check for `webOS` in `requiresCustomSubtitlesElement` [here](https://github.com/jellyfin/jellyfin-web/blob/5823cfcd54576b304522ca607e41465c4577e8f7/src/plugins/htmlVideoPlayer/plugin.js#L1138).

I am not sure about Tizen. Would it make sense to add it to `requiresCustomSubtitlesElement`? That seems like something that may have been missed.

I did update `renderSubtitlesWithCustomElement` to also handle a secondary subtitle track.

is343 replied:
I just pushed changes for the other things you pointed out. I am totally open to adding more conditions for this flag.

I assumed that the custom rendered subtitles was covering everything that doesn't support the native html ones. But now that you mention Tizen, I do see that Tizen is not in the `requiresCustomSubtitlesElement` check.

I am not too familiar with intricacies Tizen/webOS rendering since I don't have a smart tv that I can test this with.

What limits/conditions would you suggest I add?

dmitrylyzo replied:
I checked in webOS emulator (1.2, 3.0 and 4.0). It seems to support secondary subtitles natively, but it doesn't accept CSS style for pseudo elements (`::cue`, `::-webkit-media-text-track-container` and `::-webkit-media-text-track-display`). That's probably why the native element is not used.

Real Tizen 3 and Tizen 4 support secondary subtitles natively, and CSS for pseudo elements works.

So I suppose we can leave `SecondarySubtitles` feature unconditional. Anyway, we can use a custom element if it doesn't work somewhere.

is343 replied:
Okay, sounds good. Thank you for checking and confirming that.

is343 commented on src/controllers/playback/video/index.js:
```diff
@@ -998,18 +1046,37 @@ import { setBackdropTransparency, TRANSPARENCY_LEVEL } from '../../../components
                 actionsheet.show({
                     title: globalize.translate('Subtitles'),
                     items: menuItems,
+                    resolveOnClick: true,
```
Oh I wasn't aware of this. Thanks for the heads up, we can get rid of this.

is343 commented:
> I think we should only allow a combination of text-based subtitles. At the moment, ASS+ASS obviously doesn't work - it requires a second instance of JSO, but makes no sense since they will overlap. SRT+ASS or ASS+SRT are possible, but sometimes ASS is turned off when changing the subtitles. But again, this will look weird.

I definitely agree that any embedded subtitles should not be used with text based ones.
Right now I'm getting the valid secondary subtitle list is generated with `secondarySubtitleTracks` in `playbackmanager.js` by filtering for only `External` ones. 

This was me covering the problem for secondary subtitles, but for some reason I completely forgot about the scenario of if the primary ones are also embedded. Thanks for pointing this out @dmitrylyzo.

So, in this case I think I can also make a check to only allow the secondary subtitle menu if the current subtitle is also `External`. I would also probably need a check when the user resumes playback to ensure that the primary is sub is still a valid pair for a secondary.

is343 commented:
> SSA/ASS can be external. In fact, we currently only support external ones and render them using JavascriptSubtitlesOctopus. I think you could filter secondary subtitles by [`Codec`](https://github.com/jellyfin/jellyfin-web/blob/ba4a1f2e7e1ca253469c06db565f14ec70470550/src/plugins/htmlVideoPlayer/plugin.js#L1403-L1404) in addition to `DeliveryMethod`.
> ...

Ohhh okay. Sorry that initially flew over my head. I completely did not think of the complexities of rendering SSA/ASS subtitles and also the risk of subtitle overlap.

I understand what you mean now. Thank you for providing more context on that.

I refactored the method logic for checking for both valid secondary subtitle and primary subtitle pair.
It's now checking to make sure the codec format is not `SSA`, not `ASS`, and that the delivery method is `External`

is343 commented on src/plugins/htmlVideoPlayer/plugin.js:
```diff
@@ -269,6 +284,14 @@ function tryRemoveElement(elem) {
          * @type {any | undefined}
          */
         _currentPlayOptions;
+        /**
+         * @type {number}
+         */
+        _PRIMARY_TEXT_TRACK_INDEX = 0;
+        /**
+         * @type {number}
+         */
+        _SECONDARY_TEXT_TRACK_INDEX = 1;
```
That's correct. I wasn't sure where the best place was to put these sorts of constants. I didn't see any convention elsewhere and I didn't want to just plop consts at the top of the file all willy-nilly.
If you say it's fine, I can definitely go ahead and do that.

is343 commented on src/plugins/htmlVideoPlayer/plugin.js:
```diff
@@ -552,23 +586,65 @@ function tryRemoveElement(elem) {
+         * remain next to the new tracks until they reach the new offset's instance of the track.
+         */
+        requiresHidingActiveCuesOnOffsetChange() {
+            if (browser.firefox) {
+                return true;
+            }
+            return false;
```
Oh yeah, thanks. I originally wrote it that way because I wasn't sure if there were going to be other cases.

dmitrylyzo commented on src/plugins/htmlVideoPlayer/plugin.js:
```diff
@@ -1157,16 +1323,30 @@ function tryRemoveElement(elem) {
+                    secondarySubtitlesElement.classList.add('videoSecondarySubtitlesInner');
+                    subtitlesContainer.prepend(secondarySubtitlesElement);
+                    this.#videoSecondarySubtitlesElem = secondarySubtitlesElement;
+                    this.setSubtitleAppearance(subtitlesContainer, this.#videoSecondarySubtitlesElem);
```
`setSubtitleAppearance` requires some fix.
Force using of custom element and try to change `Vertical Position` in the Subtitles Settings to a positive value (3).
In case of positive position `videoSecondarySubtitlesInner` should go after `videoSubtitlesInner`.

is343 replied:
This is a really great point. I had not at all considered the vertical position setting. I will swap ordering depending on a positive or negative value.

dmitrylyzo commented on src/plugins/htmlVideoPlayer/style.scss:
```diff
@@ -74,6 +74,15 @@ video[controls]::-webkit-media-controls {
+    margin: auto;
+    display: block;
+    min-height: 0 !important;
+    margin-bottom: 0.5em !important;
```
see also the post about `setSubtitleAppearance`.
```suggestion
    margin-bottom: 0.5em !important;
    margin-top: 0.5em !important;
```

dmitrylyzo reviewed (commented):
I can still combine SRT as secondary and ASS as primary.
1. Select SRT as primary.
2. Select SRT as secondary.
3. Select ASS as primary.

Shouldn't we disable the secondary if non-eligible primary is selected?

is343 commented on src/components/playback/playbackmanager.js:
```diff
@@ -889,6 +913,20 @@ class PlaybackManager {
             return getSubtitleStream(player, index);
         }
 
+        function getCurrentSecondarySubtitleStream(player) {
+            if (!player) {
+                throw new Error('player cannot be null');
+            }
+
+            const index = getPlayerData(player).secondarySubtitleStreamIndex;
+
+            if (index == null || index === -1) {
+                return null;
+            }
+
+            return getSubtitleStream(player, index);
+        }
```
This is a great question and call out.
I originally did not want to clutter up the current method logic, so I created smaller single-purpose methods.
I'm open to do whatever fits the style of the codebase best. @thornbill, how would you suggest I organize the secondary subtitle specific logic?

thornbill replied:
Sorry for the delay. I've been a bit busy with other things for the past week. :slightly_smiling_face: 

I think I would generally prefer not splitting these out into separate functions. Just passing a boolean as needed sounds like a good idea to me.

is343 replied:
Cool, I went ahead and added an extra boolean to `getCurrentSubtitleStream` and removed this redundant method.

is343 commented on src/components/playback/playbackmanager.js:
```diff
@@ -1531,9 +1587,51 @@ class PlaybackManager {
+            const addingStream = !currentStream && newStream;
+            // Secondary subtitles are currently only handled client side
+            // Changes to the server code are required before we can handle other delivery methods
+            if (!clearingStream && (changingStream || addingStream) && getDeliveryMethod(newStream) !== 'External') {
```
You're absolutely right. This can just be simplified to `newStream` I wrote all the cases before I knew exactly what the limitations would be for client-side only work. I intended to go back and update them as I understood under what conditions we could display a secondary track.
It turns out that it is only for new streams for the scope of this PR!
I'll simplify and clean this up!

is343 commented on src/controllers/playback/video/index.js:
```diff
@@ -973,9 +973,57 @@ import { setBackdropTransparency, TRANSPARENCY_LEVEL } from '../../../components
 
+        function showSecondarySubtitlesMenu(actionsheet, positionTo) {
+            const player = currentPlayer;
+            if (!playbackManager.playerHasSecondarySubtitleSupport(player)) return;
```
Yeah, since I am still pretty new to the codebase, I didn't know all of the edge cases we could run into, so I wanted to be as safe as possible

is343 commented on src/controllers/playback/video/index.js:
```diff
@@ -998,6 +1046,29 @@ import { setBackdropTransparency, TRANSPARENCY_LEVEL } from '../../../components
+                streams.length > 1 &&
+                secondaryStreams.length > 0 &&
+                currentIndex !== -1 &&
+                playbackManager.isSubtitleStreamExternal(currentIndex, player)
```
Yup, you're right. This is also the cause of a bug that was was causing the secondary subtitle menu option to be visible when playing an `ssa/ass` track.

is343 commented on src/plugins/htmlVideoPlayer/plugin.js:
```diff
@@ -533,12 +577,15 @@ function tryRemoveElement(elem) {
-                } else if (this.#currentTrackEvents) {
-                    this.setTrackEventsSubtitleOffset(this.#currentTrackEvents, offsetValue);
+                if (trackElements.length > 0) {
+                    trackElements.forEach(function (trackElement, index) {
```
Sure thing, will use an arrow function here.

is343 commented on src/plugins/htmlVideoPlayer/plugin.js:
```diff
@@ -548,10 +595,21 @@ function tryRemoveElement(elem) {
             let relativeOffset = offsetValue;
             const newTrackOffset = offsetValue;
-            if (this.#currentTrackOffset) {
+            if (this.#currentTrackOffset && !skipRelativeOffset) {
```
This is a really great find! Thank you for pointing out this bug! You are correct, I was able to solve it by storing a separate value for secondary instead of skipping that relative offset calculation.

is343 commented:
I was away for awhile, so I was not able to respond to this feedback in a timely manner.

A real great thanks to @dmitrylyzo for really providing some excellent feedback and helping find bugs that I had not considered! I really truly do appreciate having someone wayyy more knowledgeable with the codebase and project taking a look and helping me improve my code!

@thornbill, @dmitrylyzo made a good comment about me creating several secondary subtitle helper functions/methods. Should we keep everything separate like how I currently have it? Or, should I be putting this logic directly in the existing methods?

My thinking was that I did not want to clutter the existing code, and I also wanted to keep the logic separate and tidy with their own related function names.

is343 commented:
> I can still combine SRT as secondary and ASS as primary.
>
>     1. Select SRT as primary.
>     2. Select SRT as secondary.
>     3. Select ASS as primary.
> 
> Shouldn't we disable the secondary if non-eligible primary is selected?

Yes, it should have been removing the secondary one when you changed back to ASS for the primary. I added a check in [setSubtitleStreamIndex](https://github.com/jellyfin/jellyfin-web/blob/d8d5cbf3c5b90a694c9a37544e3bfd3de80fa50a/src/components/playback/playbackmanager.js#L1592) that fixes this.

is343 commented on src/components/playback/playbackmanager.js:
```diff
@@ -1531,9 +1587,49 @@ class PlaybackManager {
+            // Also disable secondary subtitles when disabling the primary
+            // subtitles, or if it doesn't support a secondary pair
+            if (selectedTrackElementIndex === -1 || !self.trackHasSecondarySubtitleSupport(newStream)) {
+                self.setSecondarySubtitleStreamIndex(selectedTrackElementIndex);
```
Yup, that's a good point now that there are two options

is343 commented on src/controllers/playback/video/index.js:
```diff
@@ -998,6 +1046,29 @@ import { setBackdropTransparency, TRANSPARENCY_LEVEL } from '../../../components
+            streams.length > 1 &&
+            secondaryStreams.length > 0 &&
+            currentIndex !== -1 &&
+            playbackManager.trackHasSecondarySubtitleSupport(streams[currentIndex + 1], player);
```
Ahh, this is another thing that I had not considered. I'll go ahead and make `getSubtitleStream` accessible and just use that.

is343 commented on src/plugins/htmlVideoPlayer/plugin.js:
```diff
@@ -1157,16 +1334,43 @@ function tryRemoveElement(elem) {
-        renderSubtitlesWithCustomElement(videoElement, track, item) {
+        renderSubtitlesWithCustomElement(videoElement, track, item, targetTextTrackIndex) {
+            if (this.#subtitleVerticalPosition == null) {
+                import('../../scripts/settings/userSettings').then((userSettings) => {
+                    const subtitleAppearance = userSettings.getSubtitleAppearanceSettings();
+                    this.#subtitleVerticalPosition = subtitleAppearance.verticalPosition;
+                    this.#subtitleVerticalPosition = parseInt(subtitleAppearance.verticalPosition, 10);
+                });
```
Aha, I had not considered what would happen in the event of a race condition. It probably wouldn't have broken too much, other than the subtitle order.
But you're right, it is much safer for us to wait for the results with `Promise.all`

is343 commented on src/plugins/htmlVideoPlayer/style.scss:
```diff
@@ -74,6 +74,16 @@ video[controls]::-webkit-media-controls {
+    max-width: 70%;
+    background-color: rgba(0, 0, 0, 0.8);
+    margin: auto;
+    display: block;
```
This is some great inside knowledge that I wouldn't have found out otherwise.
I went ahead and made these changes. However, I was a bit confused because this diff places the padding options inside `video[controls]::-webkit-media-controls`, when they are actually  in `.videoSubtitles`.

Either way the subtitles look good with flexbox.

dmitrylyzo replied:
> when they are actually in `.videoSubtitles`.

This is right - it should go into `.videoSubtitles` when using `patch -p1 < patchfile`.

`@@ -65,20 +65,19 @@ video[controls]::-webkit-media-controls {` was genered by `git diff` :man_shrugging:

is343 commented on src/plugins/htmlVideoPlayer/plugin.js:
```diff
@@ -1157,16 +1334,43 @@ function tryRemoveElement(elem) {
+                    secondarySubtitlesElement.classList.add('videoSecondarySubtitlesInner');
+                    // determine the order of the subtitles
+                    if (this.#subtitleVerticalPosition < 0) {
+                        subtitlesContainer.prepend(secondarySubtitlesElement);
```
Thanks, this is another thing that I wasn't aware of. I'm glad you caught this.

is343 commented on src/plugins/htmlVideoPlayer/plugin.js:
```diff
@@ -533,12 +586,15 @@ function tryRemoveElement(elem) {
-                    this.setTextTrackSubtitleOffset(trackElement, offsetValue);
-                } else if (this.#currentTrackEvents) {
-                    this.setTrackEventsSubtitleOffset(this.#currentTrackEvents, offsetValue);
+                if (trackElements.length > 0) {
```
Yup, that's a good callout

is343 commented:
@thornbill Please let me know what you think of this whenever you get a chance to check it out again.
The code is much better thanks to all the great feedback from @dmitrylyzo.
I would love to get your eyes on it once more too.

is343 commented:
@thornbill any idea when this might get a final look through? I'd love to get this feature added.

thornbill commented:
Hey @is343 thanks for your patience. I was out for a few weeks due to illness and holidays, but I'm slowly starting to make my way through the backlog of open PRs now. The bigger PRs can take a bit of time to get through, but I'll try to look over this sometime in the next week.

is343 commented:
Thanks for the update @thornbill. I'm sorry to hear that you were ill. I hope you are now doing well. I did not mean to sound like I was rushing you, just wanted to see if there were any updates. It's a relief to know that it's at least in the pipeline down the line!

thornbill commented on src/components/playback/playbackmanager.js:
```diff
@@ -1463,6 +1487,24 @@ class PlaybackManager {
+                    return player.getSecondarySubtitleStreamIndex();
+                }
+            } catch (e) {
+                console.error(`Failed to get secondary stream index: ${e}`);
```
Very minor, but for debugging I like to include the filename in logs and the actual error should just be passed as a second param.

```suggestion
                console.error('[playbackmanager] Failed to get secondary stream index:', e);
```

is343 replied:
Yup sounds good!

thornbill reviewed (commented):
I only have one minor comment regarding logging really. I did some quick testing and it seems to work as expected. :tada: Unless @dmitrylyzo has some additional feedback, I'd like to get this merged so people can start testing it in unstable.

is343 commented:
> I only have one minor comment regarding logging really. I did some quick testing and it seems to work as expected. 🎉 Unless @dmitrylyzo has some additional feedback, I'd like to get this merged so people can start testing it in unstable.

Thanks for checking it out! I went ahead and committed your logging suggestion and rebased. Hopefully it should be good to go!

is343 commented on src/components/playback/playbackmanager.js:
```diff
@@ -1522,9 +1564,49 @@ class PlaybackManager {
+                return;
+            }
+
+            getPlayerData(player).secondarySubtitleStreamIndex = index;
+
+            try {
+                player.setSecondarySubtitleStreamIndex(index);
```
That's a good callout

is343 commented on src/components/playback/playbackmanager.js:
```diff
@@ -1463,6 +1487,24 @@ class PlaybackManager {
         };
 
+        self.getSecondarySubtitleStreamIndex = function (player) {
+            player = player || self._currentPlayer;
+
+            try {
+                if (player && !enableLocalPlaylistManagement(player)) {
+                    return player.getSecondarySubtitleStreamIndex();
+                }
+            } catch (e) {
+                console.error('[playbackmanager] Failed to get secondary stream index:', e);
+            }
+
+            if (!player) {
+                throw new Error('player cannot be null');
+            }
+
+            return getPlayerData(player).secondarySubtitleStreamIndex;
```
Yes definitely. Since I'm new to the project, I was trying to stick to existing code style and keep my changes as minimal as I could. Going forward, I will try to go ahead and clean things up as much as I can.

is343 commented on src/controllers/playback/video/index.js:
```diff
@@ -1013,6 +1061,29 @@ import { setBackdropTransparency, TRANSPARENCY_LEVEL } from '../../../components
+            streams.length > 1 &&
+            secondaryStreams.length > 0 &&
+            currentIndex !== -1 &&
+            playbackManager.trackHasSecondarySubtitleSupport(playbackManager.getSubtitleStream(player, currentIndex), player);
```
Thanks for this. Is there any lint or prettier check that I can run to catch these? I don't think the current `lint` script catches stuff like this.

dmitrylyzo replied:
We can add later (need to fix other places):
```diff
diff --git a/.eslintrc.js b/.eslintrc.js
index 53dd01ff2..362b3abde 100644
--- a/.eslintrc.js
+++ b/.eslintrc.js
@@ -64,6 +64,7 @@ module.exports = {
         'no-void': ['error', { 'allowAsStatement': true }],
         'no-warning-comments': ['warn', { 'terms': ['fixme', 'hack', 'xxx'] }],
         'one-var': ['error', 'never'],
+        'operator-linebreak': ['error', 'before'],
         'padded-blocks': ['error', 'never'],
         'prefer-const': ['error', { 'destructuring': 'all' }],
         'quotes': ['error', 'single', { 'avoidEscape': true, 'allowTemplateLiterals': false }],
```
But I didn't find how to check indentation after line-break on operator.

thornbill replied:
The indentation check is still disabled in this file. At some point we do need to go through and fix those. 😅

is343 commented on src/plugins/htmlVideoPlayer/plugin.js:
```diff
@@ -549,23 +601,63 @@ function tryRemoveElement(elem) {
             return relativeOffset;
         }
 
+        /**
+         * @private
+         * These browsers will not clear the existing active cue when setting an offset
+         * for native TextTracks.
+         * Any previous text tracks that are on the screen when the offset changes will
+         * remain next to the new tracks until they reach the new offset's instance of the track.
+         */
+        requiresHidingActiveCuesOnOffsetChange() {
+            return !!browser.firefox;
+        }
+
+        /**
+         * @private
+         */
+        hideTextTrackActiveCues(currentTrack) {
+            if (currentTrack.activeCues) {
+                Array.from(currentTrack.activeCues).forEach((cue) => {
+                    cue.text = '';
+                });
+            }
+        }
```
@dmitrylyzo @thornbill I tried this again and it actually might not be as big of a problem as I originally thought.
I cannot reproduce this in firefox with my prod 10.8.9 on my home server. But I can confirm that I *can* still reproduce this in firefox (chrome is fine) for the current `master` branch (10.9.0) debug build via `npm start`.

The screenshot below was from me quickly changing the offset multiple times. You can see that the track cues have stacked up and are stuck. Again, this seems to not happen in the prod build. 
It only seems to happen on firefox when served via the `start` script.

So maybe this fix is only actually relevant when debugging in the first place.

<img width="616" alt="Screen Shot 2023-02-09 at 7 36 08 PM" src="https://user-images.githubusercontent.com/30599893/217972440-c09c08bf-22b6-4a95-a6c8-8ef771188030.png">

dmitrylyzo replied:
Hmm, this issue occurs even with the current fix.
I tried changing `track.mode` to `disabled`, then `showing` to remove the extra lines.

I think that if it happens in `devel` and doesn't happen in `prod`, it's just luck. :smile:

is343 replied:
Are you able to reproduce it consistently even with this fix? I'm not able to reproduce it at all still.

Does disabling then enabling fix it for you?
Like this?
```
hideTextTrackActiveCues(currentTrack) {
      currentTrack.mode = 'disabled';
      currentTrack.mode = 'showing';
}
```

is343 replied:
Disabling then showing like I mentioned above doesn't seem to work for me. I am able to reproduce the error after making the change I mentioned above.

The initial fix of clearing the cue with empty strings still works for me and I cannot reproduce the issue when the cues are blanked out

dmitrylyzo replied:
> Are you able to reproduce it consistently even with this fix?

Yes. I just open the slider and click +2..3 seconds. Clicking back (0..+1 seconds) clears them though.

Clearing them after an offset (move `hideTextTrackActiveCues` below the offset block) works, but they are not displayed when paused.

:warning: This fix seems to corrupt the cues. Subtitles are not displayed when rewinding.

> Disabling then showing like I mentioned above doesn't seem to work for me.

I had to change it like this because `currentTrack.cues` became `null` after disabling the track:
```diff
diff --git a/src/plugins/htmlVideoPlayer/plugin.js b/src/plugins/htmlVideoPlayer/plugin.js
index 700eedd14..a86b32d4f 100644
--- a/src/plugins/htmlVideoPlayer/plugin.js
+++ b/src/plugins/htmlVideoPlayer/plugin.js
@@ -639,11 +639,11 @@ function tryRemoveElement(elem) {
          * @private
          */
         hideTextTrackActiveCues(currentTrack) {
-            if (currentTrack.activeCues) {
-                Array.from(currentTrack.activeCues).forEach((cue) => {
-                    cue.text = '';
-                });
-            }
+            currentTrack.mode = 'disabled';
+
+            setTimeout(() => {
+                currentTrack.mode = 'showing';
+            }, 0);
         }
 
         /**
@@ -655,14 +655,15 @@ function tryRemoveElement(elem) {
                 if (offsetValue === 0) {
                     return;
                 }
-                if (this.requiresHidingActiveCuesOnOffsetChange()) {
-                    this.hideTextTrackActiveCues(currentTrack);
-                }
                 Array.from(currentTrack.cues)
                     .forEach(function (cue) {
                         cue.startTime -= offsetValue;
                         cue.endTime -= offsetValue;
                     });
+
+                if (this.requiresHidingActiveCuesOnOffsetChange()) {
+                    this.hideTextTrackActiveCues(currentTrack);
+                }
             }
         }
 
```
But the timeout doesn't allow the subtitle offset slider to change the offset when rapidly clicking - it falls here because all tracks are disabled (I'm testing with one subtitle track):
https://github.com/jellyfin/jellyfin-web/blob/276a59191bdc0715e61613e22f2f647d22bc34e3/src/plugins/htmlVideoPlayer/plugin.js#L596

is343 replied:
I also tried with a 0ms timeout and was running into the same issue. 

Hmmm, I will also look at this a bit more.

Since this is trying to fix an issue that is currently unrelated to the dual subtitles and a bit out of scope of this PR, do you think I should just remove this (not fully working) fix from this PR and create a new separate one specifically for this issue?

dmitrylyzo replied:
> Since this is trying to fix an issue that is currently unrelated to the dual subtitles and a bit out of scope of this PR, do you think I should just remove this (not fully working) fix from this PR and create a new separate one specifically for this issue?

Yes, I think it would be better to make another PR for that. Moreover, we can try to fix this in 10.8 (if possible).

is343 replied:
I've opened a draft PR #4349. I will continue my work on this bug there.

dmitrylyzo commented on src/plugins/htmlVideoPlayer/plugin.js:
```diff
@@ -1155,16 +1327,39 @@ function tryRemoveElement(elem) {
+                    const subtitlesElement = document.createElement('div');
+                    subtitlesElement.classList.add('videoSubtitlesInner');
+                    subtitlesContainer.appendChild(subtitlesElement);
                     this.#videoSubtitlesElem = subtitlesContainer.querySelector('.videoSubtitlesInner');
```
```suggestion
                    this.#videoSubtitlesElem = subtitlesElement;
```

is343 commented:
@dmitrylyzo @thornbill Just a heads up to you both, I noticed that direct play now seems to be broken in this branch.
The break seems to be happening in the html plugin file. Reverting **only** that file to the current master seems to fix it. I'm going through commits and diffs to see what has changed in master that broke this. 

I'll update you once I figure it out.

![SCR-20230210-gc4](https://user-images.githubusercontent.com/30599893/218152146-8f0eca15-e856-42db-a0cb-b0e95206607d.png)

is343 commented on src/plugins/htmlVideoPlayer/plugin.js:
```diff
@@ -480,19 +480,19 @@ function tryRemoveElement(elem) {
 
             if (!secondaryTrackValid) {
                 currentSecondaryTrackIndex = -1;
-                playbackManager.setSecondarySubtitleStreamIndex(currentSecondaryTrackIndex);
```
@thornbill @dmitrylyzo to update regarding that regression mentioned earlier, it was caused by these calls to `playbackManager`.

They were originally using the default fallback player defined in those methods, but I guess the timing has now changed. These were coming up as `undefined` in the methods, but they were never actually entering into the logic that throws an error for some reason. 
```
if (!player) {
    throw new Error('player cannot be null');
}
```

Maybe there was some sort of race condition where the HLS player would start before it reached that point? I'm not sure.

Either way, the fix was to just pass `this` into those methods, and it is now working as it should.

![SCR-20230210-jek](https://user-images.githubusercontent.com/30599893/218175319-26767881-b838-4d47-ab10-9efeb366b3a7.png)

is343 commented on src/plugins/htmlVideoPlayer/plugin.js:
```diff
@@ -480,19 +480,19 @@ function tryRemoveElement(elem) {
             if (!secondaryTrackValid) {
                 currentSecondaryTrackIndex = -1;
-                playbackManager.setSecondarySubtitleStreamIndex(currentSecondaryTrackIndex);
+                playbackManager.setSecondarySubtitleStreamIndex(currentSecondaryTrackIndex, this);
```
@thornbill @dmitrylyzo to update regarding that regression mentioned earlier, it was caused by these calls to `playbackManager`.

They were originally using the default fallback player defined in those methods, but I guess the timing has now changed. These were coming up as `undefined` in the methods, but they were never actually entering into the logic that throws an error for some reason. 
```
if (!player) {
    throw new Error('player cannot be null');
}
```

Maybe there was some sort of race condition where the HLS player would start before it reached that point? I'm not sure.

Either way, the fix was to just pass `this` into those methods, and it is now working as it should.

![SCR-20230210-jek](https://user-images.githubusercontent.com/30599893/218175319-26767881-b838-4d47-ab10-9efeb366b3a7.png)

dmitrylyzo replied:
@is343 I noticed that you removed this commit. Accidentally?

is343 replied:
@dmitrylyzo oh no! **Yes thank you so much for catching that!**
It was accidental.
I was debugging on a different machine when I made this change and pushed this commit.
I must have forgot to pull the changes when I went back to my original machine. Then rebased upstream master, then `push --force-with-lease` to rebase this with master.

is343 replied:
I just repushed with this commit back in

is343 commented:
I have removed the work to fix the stuck track cues for dev firefox builds into a separate draft PR #4349. This work is out of the scope of this PR.
I also fixed the previous direct playback regression bug that was found after the last rebase with master.

Hopefully we can move this into unstable so we can get more people testing this 🤞.

@thornbill @dmitrylyzo, please take another look to make sure it is good to go. Thank you both for really helping me improve the quality of this PR with your reviews. And thank you both for your patience.

dmitrylyzo commented on src/plugins/htmlVideoPlayer/plugin.js:
```diff
@@ -448,11 +470,36 @@ function tryRemoveElement(elem) {
+
+            if (!secondaryTrackValid) {
+                currentSecondaryTrackIndex = -1;
+                playbackManager.setSecondarySubtitleStreamIndex(currentSecondaryTrackIndex);
```
This should go after
https://github.com/jellyfin/jellyfin-web/blob/f4abbf3e22aec8c3743e819d3764890aba98277a/src/plugins/htmlVideoPlayer/plugin.js#L508

Strange that this issue didn't occur with `devtool: 'source-map'` in `webpack.dev.js`.

is343 replied:
Oh hmm. Okay let me take a look at this

is343 replied:
Alright, the most recent 2 commits fix these issues. One for the bug, and the other for clearing and setting the index.

There are soooo many different states being passed and saved, it is really difficult to keep track of what is doing what. I tried to set the minimum amount to allow the secondary subtitles to be carried over to the next media source when playing multiple episodes of a show.

I also decided to not keep the saved secondary track index state once you back out. As you mentioned, it was getting saved to the player state so it would continue to exist for the current session while it is loaded.
But after thinking about this it doesn't make sense to keep the index stored like that. The user can go from an episode to a movie. They might be entirely unrelated with subtitles, so the index is not useful in that case.

Since we do not have any stored secondary subtitle language settings that we can pull from, similar to how subtitle and audio languages are fetched, I don't think it is appropriate to include at the moment.

So now, the flow for auto playing secondary subtitles is:
```
user clicks on a video -> 
selects the secondary subtitle -> 
clicks to next/prev video from the player -> 
the currently active secondary subtitle goes into the current `rankStreamType` logic to find the best match
```
https://github.com/jellyfin/jellyfin-web/blob/f4abbf3e22aec8c3743e819d3764890aba98277a/src/components/playback/playbackmanager.js#L2314

is343 commented:
I also have a general question. I noticed as I was working on the last commit that the `getPlayerData` method prepares an object for state, but we never actually use it. We instead save everything to the player directly.

I tried changing it to `return state;` instead, and everything seemed to continue to work fine. I don't know if there are any serious consequences to keeping the state in the player compared to an separate state object, but as it is now, we don't use the `playerStates` for anything at the moment. It looks like it's been this way for several years.

https://github.com/jellyfin/jellyfin-web/blob/85fac85968a877679ecde866358172a824e3ab62/src/components/playback/playbackmanager.js#L1919

dmitrylyzo commented:
> I also have a general question. I noticed as I was working on the last commit that the `getPlayerData` method prepares an object for state, but we never actually use it. We instead save everything to the player directly.
> 
> I tried changing it to `return state;` instead, and everything seemed to continue to work fine. I don't know if there are any serious consequences to keeping the state in the player compared to an separate state object, but as it is now, we don't use the `playerStates` for anything at the moment. It looks like it's been this way for several years.

As you may have noticed, the current sources are like spaghetti, so the main reason they are left as is "I don't know if there are any serious consequences ...".

Ideally, PlaybackManager should mediate the playback process, and players are just wrappers for the media API. The mentioned data can be stored by the player and should be wrapped in getters so that the state is always in sync (currently, it may differ if you have made a mistake).

dmitrylyzo commented on src/plugins/htmlVideoPlayer/plugin.js:
```diff
@@ -448,18 +470,43 @@ function tryRemoveElement(elem) {
 
             this._currentPlayOptions = options;
 
+            // Get the secondary track that has been set during this watch session
+            let currentSecondaryTrackIndex = playbackManager.getSecondarySubtitleStreamIndex(this);
+
+            if (!secondaryTrackValid) {
+                currentSecondaryTrackIndex = -1;
+                playbackManager.setSecondarySubtitleStreamIndex(currentSecondaryTrackIndex, this);
+            }
+
+            this.#secondarySubtitleTrackIndexToSetOnPlaying = currentSecondaryTrackIndex == null ? -1 : currentSecondaryTrackIndex;
+            if (this.#secondarySubtitleTrackIndexToSetOnPlaying != null && this.#secondarySubtitleTrackIndexToSetOnPlaying >= 0) {
+                const initialSecondarySubtitleStream = options.mediaSource.MediaStreams[this.#secondarySubtitleTrackIndexToSetOnPlaying];
+                if (!initialSecondarySubtitleStream || initialSecondarySubtitleStream.DeliveryMethod !== 'External') {
+                    this.#secondarySubtitleTrackIndexToSetOnPlaying = -1;
+                }
+            }
```
Using `getSecondarySubtitleStreamIndex` has a side effect:
1. Run video with 2 SRTs.
2. Select both subtitles (primary and secondary).
3. Stop playing.
4. Run the same video.
5. Secondary subtitles are displayed, but they are not marked in the menu because they are cleared in `onPlaybackStarted`. :spaghetti: :man_facepalming: 

If we want to reset the secondary subtitle selection when we start a new session, we can do the following:
```suggestion
            if (secondaryTrackValid) {
                this.#secondarySubtitleTrackIndexToSetOnPlaying = options.mediaSource.DefaultSecondarySubtitleStreamIndex == null ? -1 : options.mediaSource.DefaultSecondarySubtitleStreamIndex;
                if (this.#secondarySubtitleTrackIndexToSetOnPlaying != null && this.#secondarySubtitleTrackIndexToSetOnPlaying >= 0) {
                    const initialSecondarySubtitleStream = options.mediaSource.MediaStreams[this.#secondarySubtitleTrackIndexToSetOnPlaying];
                    if (!initialSecondarySubtitleStream || initialSecondarySubtitleStream.DeliveryMethod !== 'External') {
                        this.#secondarySubtitleTrackIndexToSetOnPlaying = -1;
                        playbackManager.setSecondarySubtitleStreamIndex(-1, this);
                    }
                }
            } else {
                this.#secondarySubtitleTrackIndexToSetOnPlaying = -1;
                playbackManager.setSecondarySubtitleStreamIndex(-1, this);
            }
```
`playbackManager.setSecondarySubtitleStreamIndex(-1, this)` can probably be removed if we remove this condition:
https://github.com/jellyfin/jellyfin-web/blob/3c59efaf700622384483bd581100ae748bcbf435/src/plugins/htmlVideoPlayer/plugin.js#L901
But I am not sure if this will cause problems with playbackmanager (race).

is343 replied:
Okay, so I tried with removing the `playbackManager.setSecondarySubtitleStreamIndex(-1, this)` and it seems to be working fine. I have not run into anything strange. 

It is now functioning as I mentioned before. It will always reset the secondary subtitle to a `null/undefined` value by default. The secondary subtitle track index will only be carried over if we are going to the next/prev track from the player itself.

I think this is the best approach for now until some sort of official user secondary subtitle default language setting can be implemented.

is343 commented on src/components/playback/playbackmanager.js:
```diff
@@ -1639,6 +1724,7 @@ class PlaybackManager {
             }).then(function (deviceProfile) {
                 const audioStreamIndex = params.AudioStreamIndex == null ? getPlayerData(player).audioStreamIndex : params.AudioStreamIndex;
                 const subtitleStreamIndex = params.SubtitleStreamIndex == null ? getPlayerData(player).subtitleStreamIndex : params.SubtitleStreamIndex;
+                const secondarySubtitleStreamIndex = params.SubtitleStreamIndex == null ? getPlayerData(player).secondarySubtitleStreamIndex : params.secondarySubtitleStreamIndex;
```
I will include all of these changes in one commit

dmitrylyzo reviewed (commented):
I amended the last commit to make fewer changes.

Found another 2 bugs.

The first: AutoSet sets SSA/ASS as secondary - see suggestion.

**Steps To Reproduce**
1. Setup 2 episodes:
    1. `lang1` (SRT), `lang2` (SRT)
    3. `lang1` (SRT), `lang2` (ASS)
2. Start a queue (playlist) with both episodes.
4. Select `lang1` as the primary, `lang2` as the secondary.
5. Skip first episode.
6. `lang2` is automatically selected, but it cannot be played with `lang1` (SRT).

And the second one: `playerData.secondarySubtitleStreamIndex` is still not reset if the secondary subtitle is wrong (bug above), but it is now unnoticeable.
[This line](https://github.com/jellyfin/jellyfin-web/blob/ba47fb65e35fbde1492497a081b990af58456373/src/components/playback/playbackmanager.js#L3016) reverts it even if we call `playbackManager.setSecondarySubtitleStreamIndex(-1, this);` in `htmlVideoPlayer`.

Maybe we should leave the second bug for later, since I dunno how to fix it right now? :sweat_smile: 
[This block](https://github.com/jellyfin/jellyfin-web/blob/ba47fb65e35fbde1492497a081b990af58456373/src/components/playback/playbackmanager.js#L3013-L3021) is probably completely wrong.

is343 commented on src/components/playback/playbackmanager.js:
```diff
@@ -1522,9 +1568,48 @@ class PlaybackManager {
+
+            // Secondary subtitles are currently only handled client side
+            // Changes to the server code are required before we can handle other delivery methods
+            if (newStream && getDeliveryMethod(newStream) !== 'External') {
```
I agree, we should definitely be using that method to check if it is valid

is343 commented:
> And the second one: `playerData.secondarySubtitleStreamIndex` is still not reset if the secondary subtitle is wrong (bug above), but it is now unnoticeable. [This line](https://github.com/jellyfin/jellyfin-web/blob/ba47fb65e35fbde1492497a081b990af58456373/src/components/playback/playbackmanager.js#L3016) reverts it even if we call `playbackManager.setSecondarySubtitleStreamIndex(-1, this);` in `htmlVideoPlayer`.
> 
> Maybe we should leave the second bug for later, since I dunno how to fix it right now? 😅 [This block](https://github.com/jellyfin/jellyfin-web/blob/ba47fb65e35fbde1492497a081b990af58456373/src/components/playback/playbackmanager.js#L3013-L3021) is probably completely wrong.

I think if we also check if the "best track index" is a valid secondary track inside of `rankStreamType`, we might be able to reset it properly. We are already passing a `isSecondarySubtitle` flag, so we should be able to also check if it's valid.
Let me try this and I will let you know

dmitrylyzo commented:
> I think if we also check if the "best track index" is a valid secondary track inside of `rankStreamType`, we might be able to reset it properly. We are already passing a `isSecondarySubtitle` flag, so we should be able to also check if it's valid.
> Let me try this and I will let you know

You may need to check if both subtitles are compatible for secondary (in `autoSetNextTracks`) and reset `DefaultSecondarySubtitleStreamIndex` if they are not.
This will probably work (as a workaround), but the problem is that subtitle tracks are stored in playbackManager (playerData) and the player (`#customTrackIndex` and `#customSecondaryTrackIndex`) and should be always in sync. Storing and managing them in one place should eliminate the synchronization problem. But this is for the next PR :wink:

is343 commented:
> You may need to check if both subtitles are compatible for secondary (in `autoSetNextTracks`) and reset `DefaultSecondarySubtitleStreamIndex` if they are not. This will probably work (as a workaround), but the problem is that subtitle tracks are stored in playbackManager (playerData) and the player (`#customTrackIndex` and `#customSecondaryTrackIndex`) and should be always in sync. Storing and managing them in one place should eliminate the synchronization problem. But this is for the next PR 😉

Yes I agree. The way we have to manage two states is not ideal. We will definitely need to followup on this.
But for now, I think I have it working. It will now check the secondary track and only carry the best match index if it is a valid secondary track. When carrying over the values after we do the ranking, we will also ensure the primary track is valid before setting the secondary track as well.

is343 commented on src/components/playback/playbackmanager.js:
```diff
@@ -2910,9 +3017,15 @@ class PlaybackManager {
             if (mediaSource) {
                 playerData.audioStreamIndex = mediaSource.DefaultAudioStreamIndex;
                 playerData.subtitleStreamIndex = mediaSource.DefaultSubtitleStreamIndex;
+                if (self.trackHasSecondarySubtitleSupport(mediaSource.MediaStreams[mediaSource.DefaultSubtitleStreamIndex])) {
```
That would definitely clean things up. What about if the `rankStreamType` best choice happens to be an unsupported secondary stream option? It's unlikely to get enough points because it would not get any points for the matching codec, but it's still possible. 
If it chooses an unsupported track, then we would get the same issue where it will not show in the player, but the index would still be saved in the player state.

https://github.com/jellyfin/jellyfin-web/blob/e1521f9a125c38444afecfe089cbb2140e1821ab/src/components/playback/playbackmanager.js#L2368

Maybe you are correct in that maybe we should consider this as work for a separate PR

dmitrylyzo replied:
> What about if the `rankStreamType` best choice happens to be an unsupported secondary stream option? It's unlikely to get enough points because it would not get any points for the matching codec, but it's still possible.

The proposed code checks such a situation.
`rankStreamType` and `autoSetNextTracks` are private, so it is safe to do it not in them.

dmitrylyzo reviewed (approved):
_Moved `streamInfo` down._

The playerData bug (inconsistency) requires refactoring playbackManager and players, so not now.

is343 commented:
@thornbill Alright, we figured out the kinks. Please give this a final look when you get the chance. :raised_hands:

WarlockFish commented:
why i cant see the selecting secondary subtitles button？

is343 commented:
> why i cant see the selecting secondary subtitles button？

Hi, this feature is going to be included in the version `10.9` release.

Right now, the only way to test this out is to try one of the nightly/unstable builds.

If you do try to update to one of the 10.9 unstable builds, please be aware that you won't be able to downgrade back to the 10.8 builds.

Soulxy1993 commented:
@is343 
Is this included in the version 10.9.1?
I updated to the newest version but can't find the function.

thornbill commented:
If you have a question, pick a method from https://jellyfin.org/contact to ask it... don't comment on old PRs.
